finality-aleph/src/base_protocol/service.rs

```diff
@@ -102,16 +122,75 @@ where
+                        debug!(target: LOG_TARGET, "Rejecting incoming substream: {}.", e);
+                        ValidationResult::Reject
+                    }
+                };
+                if result_tx.send(result).is_err() {
+                    debug!(
+                        target: LOG_TARGET,
+                        "Failed to send response to inbound substream validation request."
+                    );
+                }
+            }
+            NotificationStreamOpened {
+                peer,
+                handshake,
+                direction,
+                negotiated_fallback: _,
+            } => match self.handler.on_peer_connect(peer, handshake, direction) {
+                Ok(()) => {
+                    let multiaddress: Multiaddr =
+                        iter::once(MultiaddressProtocol::P2p(peer.into())).collect();
+                    trace!(target: LOG_TARGET, "Connected event from address {:?}.", multiaddress);
+                    for name in &self.protocol_names {
+                        if let Err(e) = self.network.add_peers_to_reserved_set(
+                            name.clone(),
+                            iter::once(multiaddress.clone()).collect(),
+                        ) {
+                            error!(target: LOG_TARGET, "Adding peer to the {} reserved set failed: {}.", name, e);
+                        }
+                    }
+                }
+                Err(e) => debug!(target:LOG_TARGET, "Failed to accept connection: {}.", e),
+            },
+            NotificationStreamClosed { peer } => {
+                trace!(target: LOG_TARGET, "Disconnected event for peer {:?}", peer);
+                let addresses: Vec<_> = iter::once(peer).collect();
+                for name in &self.protocol_names {
+                    if let Err(e) = self
+                        .network
+                        .remove_peers_from_reserved_set(name.clone(), addresses.clone())
+                    {
+                        warn!(target: LOG_TARGET, "Removing peer from the {} reserved set failed: {}", name, e)
+                    }
+                }
+            }
+            NotificationReceived { peer, .. } => {
+                debug!(target: LOG_TARGET, "Received unexpected message in the base protocol from {}.", peer)
+            }
+        }
+    }
+
     /// Run the service managing the base protocol.
     pub async fn run(mut self) -> Result<(), Error> {
         use Error::*;
         loop {
-            let command = self
-                .commands_from_user
-                .next()
-                .await
-                .ok_or(NoIncomingCommands)?;
-            self.handle_command(command);
+            tokio::select! {
+                command = self.commands_from_user.next() => self.handle_command(command.ok_or(NoIncomingCommands)?),
+                event = self.events_from_network.next_event() => self.handle_network_event(event.ok_or(NoNetworkEvents)?),
```

Comment:
Likely the answer is "YES", but let me ask to make sure: is breaking the select loop in case of an error in one of the branches correct?

Reply:
If any of the streams returns `None` we are definitely done, so yes.